src/main/java/org/beanpod/switchboard/controller/DeviceController.java

```diff
@@ -10,65 +11,85 @@
-import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.PutMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
 @Slf4j
 @RestController
-@RequestMapping("/device")
 @RequiredArgsConstructor
-public class DeviceController {
-
+public class DeviceController implements DeviceApi {
+  public static final String UNKNOWN_ERROR_MESSAGE = "Unknown error in Device Controller";
   private final DeviceDaoImpl service;
   private final DeviceMapper deviceMapper;
+  private final HttpServletRequest request;
 
-  @GetMapping
-  public List<DeviceDTO> retrieveAllDevices() {
-    return (deviceMapper.toDeviceDTOs(service.getDevices()));
+  @Override
+  public ResponseEntity<List<DeviceModel>> retrieveAllDevices() {
+    return Optional.of(service.getDevices())
+        .map(deviceMapper::toDeviceDTOs)
+        .map(deviceMapper::toDeviceModels)
+        .map(ResponseEntity::ok)
+        .orElseThrow(this::getUnknownException);
   }
 
-  @GetMapping("/{serialNumber}")
-  public ResponseEntity<DeviceDTO> retrieveDevice(@PathVariable String serialNumber) {
-    Optional<DeviceDTO> device = service.findDevice(serialNumber);
-    if (device.isEmpty()) {
-      throw new ExceptionType.DeviceNotFoundException(serialNumber);
-    }
-    return ResponseEntity.ok(device.get());
+  @Override
+  public ResponseEntity<DeviceModel> retrieveDevice(@PathVariable String serialNumber) {
+    DeviceDTO deviceDto =
+        service
+            .findDevice(serialNumber)
+            .orElseThrow(() -> new ExceptionType.DeviceNotFoundException(serialNumber));
+
+    return Optional.of(deviceDto)
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
+        .orElseThrow(this::getUnknownException);
   }
 
-  @PostMapping
-  public ResponseEntity<DeviceDTO> createDevice(@RequestBody @Valid DeviceDTO device) {
-    Optional<DeviceDTO> deviceLookup = service.findDevice(device.getSerialNumber());
+  @Override
+  public ResponseEntity<DeviceModel> createDevice(@Valid CreateDeviceRequest createDeviceRequest) {
+    Optional<DeviceDTO> deviceLookup = service.findDevice(createDeviceRequest.getSerialNumber());
     if (deviceLookup.isPresent()) {
-      throw new ExceptionType.DeviceAlreadyExistsException(device.getSerialNumber());
+      throw new ExceptionType.DeviceAlreadyExistsException(createDeviceRequest.getSerialNumber());
     }
-    return ResponseEntity.ok(service.save(device));
+
+    return Optional.of(createDeviceRequest)
+        .map(createRequest -> service.createDevice(createRequest, request.getRemoteAddr()))
+        .map(deviceMapper::toDeviceModel)
+        .map(ResponseEntity::ok)
```

Comment:
This is more of something to address in the future instead of addressing in this pr.
We map the entities to dao then we map it models. why not map it to models directly in the dao package

Reply:
In my mind the Controller should convert the `model` to a `dto` then send it to the service layer. The "Create...Request" are sort of exceptions to these rules because they don't have DTOs.